### ISSUE TITLE
helm chart: kube-etcd 4001 port blocked on master node in kops

### ISSUE THREAD

**Issue by inodb** [coreos/prometheus-operator#1369]
**What did you do?**
On a kops cluster with tiller running
```
helm repo add coreos https://s3-eu-west-1.amazonaws.com/coreos-charts/stable/
helm install coreos/prometheus-operator --name prometheus-operator --namespace monitoring
helm install coreos/kube-prometheus --name kube-prometheus --namespace monitoring
```

Then checked out prometheus dashboard with
```
kubectl port-forward -n monitoring kube-prometheus-grafana-595d49c45-bt49p  3000
```
The kube-etcd service wasn't working in targets. It's because 4001 port is blocked by default by kops.

**What did you expect to see?**
kube-etcd working. Ideally it would use the https port 2379

**What did you see instead? Under which circumstances?**
kube-etcd gave an alert that it was down

**Environment**

* Kubernetes version information:
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.1", GitCommit:"d4ab47518836c750f9949b9e0d387f20fb92260b", GitTreeState:"clean", BuildDate:"2018-04-13T22:29:03Z", GoVersion:"go1.9.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-07T11:55:20Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
```

* Kubernetes cluster kind: 

kops version 1.9.0 on AWS

**Comment by gianrubio** [comment]
This is not an issue in the helm charts pleas read [this issue(https://github.com/kubernetes/kops/issues/4975#issuecomment-381055946)] to fix it. Another solution will be to monitor etcd running an [extra prometheus](https://github.com/coreos/prometheus-operator/issues/898) instance using tolerations to schedule it in a master node. 

It'll be nice to have to etcd monitored out of the box with the helm charts, PR are always welcome.

**Comment by gianrubio** [comment]
With this `Prometheus` object you can launch a new prometheus in your cluster and pull metrics from etcd.

```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: etcd-prometheus
  namespace: monitoring
  labels:
    app: exporter-kube-etcd
spec:
  alerting:
    alertmanagers:
    - name: monitoring-alertmanager
      namespace: monitoring
      port: http
  externalUrl: http://prometheus-etc..../
  logLevel: info
  resources:
    limit:
      memory: 400Mi
      cpu: 0.5
  retention: 300h
  serviceAccountName: monitoring-prometheus
  ruleSelector:
    matchLabels:
      prometheus: monitoring
      role: alert-rules
      chart=exporter-kube-etcd
  serviceMonitorSelector:
    matchLabels:
      app: exporter-kube-etcd     
  nodeSelector:
    node-role.kubernetes.io/master: ""      
  tolerations:
    - key: node-role.kubernetes.io/master
      operator: Exists
      effect: NoSchedule  
```

**Comment by gianrubio** [comment]
I finally got an extra prometheus instance running in my master nodes, it's dedicated to pull metrics only from etcd.  It's a bit tricky to configure this extra instance and to not duplicate configuration between prometheus instances (serviceMonitorSelector and service labels conflicts 💀 🔫 ).  
Before I implement this in the helm chart I'd like to see how users behaves with this configs, so no PR for now.

This creates a new prometheus configuration that will run in the master nodes and pull metrics from etcd, it's required to use `nodeSelector` and `tolerations` so the prometheus instance will be scheduled in a master/etcd node. The `serviceMonitorSelector` shouldn't conflict to the other prometheus object otherwise prometheus instances will pick the same `serviceMonitorSelector`.

```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: etcd-prometheus
  labels:
    app: exporter-kube-etcd
    prometheus: etcd-prometheus
spec:
  resources:
    limit:
      memory: 100Mi
      cpu: 0.3
  retention: 300h
  serviceAccountName: monitoring-prometheus
  ruleSelector:
    matchLabels:
      prometheus: monitoring
      role: alert-rules
      chart: exporter-kube-etcd
  serviceMonitorSelector:
    matchLabels:
      app: exporter-kube-etcd     
  nodeSelector:
    node-role.kubernetes.io/master: ""      
  tolerations:
    - key: node-role.kubernetes.io/master
      operator: Exists
      effect: NoSchedule  
```

This serviceMonitor is created by kube-etcd, note that the label `app` should be changed so the etcd prometheus instance will pick it.

```
apiVersion: v1
items:
- apiVersion: monitoring.coreos.com/v1
  kind: ServiceMonitor
  metadata:
    labels:
      app: exporter-kube-etcd # <<< CHANGE-ME >> 
      prometheus: monitoring
      release: monitoring
    name: monitoring-exporter-kube-etcd
    namespace: monitoring
  spec:
    endpoints:
    - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
      interval: 15s
      port: http-metrics
    jobLabel: component
    namespaceSelector:
      matchNames:
      - kube-system
    selector:
      matchLabels:
        app: exporter-kube-etcd
        component: kube-etcd
kind: List

```

Then I created a service that will help the `serviceMonitor` to discover the new prometheus instance.

```
apiVersion: v1
kind: Service
metadata:
  labels:
    app: prometheus
    prometheus: etcd-prometheus
  name: etcd-prometheus
spec:
  ports:
  - name: http
    port: 9090
    protocol: TCP
    targetPort: 9090
  selector:
    app: prometheus
    prometheus: etcd-prometheus
```

The serviceMonitor bellow it's not required, however when you want to run in federation could be helpful. 

```
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    app: prometheus
    prometheus: monitoring
    release: monitoring
  name: prometheus-etcd
  namespace: monitoring
spec:
  endpoints:
  - honor_labels: true
    interval: 30s
    params:
      match[]:
      - '{job="kube-etcd"}'
      - '{__name__=~"job:.*"}'
    path: /federate
    port: http
  namespaceSelector:
    matchNames:
    - monitoring
  selector:
    matchLabels:
      prometheus: etcd-prometheus
```

# Prometheus-etcd
After all this configuration you should have a new pod named `monitoring prometheus-etcd-prometheus-0`, use the port forward to connect to `k port-forward   -n monitoring prometheus-etcd-prometheus-0   9090` . Because the prometheus instances are running in federation, you will see the metrics from etcd in the "master" prometheus instance

![screen shot 2018-06-13 at 09 51 06](https://user-images.githubusercontent.com/5694860/41337377-5758973e-6eef-11e8-98c6-2103815f38f5.png)

## Querying prometheus etcd

![screen shot 2018-06-13 at 10 15 09](https://user-images.githubusercontent.com/5694860/41338521-b8895f90-6ef2-11e8-89df-d2b7dcde0b56.png)

# Master prometheus

<img width="589" alt="screen shot 2018-06-13 at 10 13 32" src="https://user-images.githubusercontent.com/5694860/41338469-945c4fce-6ef2-11e8-9390-da43e04000f4.png">

## Querying the master prometheus 
![screen shot 2018-06-13 at 10 14 13](https://user-images.githubusercontent.com/5694860/41338471-9476ca70-6ef2-11e8-9659-1e00af57da59.png)

Notes: Because I'm using this new prometheus etcd in federation mode the alerts are not configured in this instance,  the "master" prometheus will pick the serviceMonitor `prometheus-etcd`.

**Comment by FuzzOli87** [comment]
@gianrubio I deployed all this through kube-prometheus and I don't currently see a way to change the app label from the top level kube-prometheus values. I also can't tell kube-prometheus not to install kube-etcd. Do you suggest just creating double configurations?

**Comment by jhohertz** [comment]
Just in case anyone else is struggling with this with the current helm charts, see this gist for an easy solution. (Disable the helm chart's etcd monitors, and apply that yaml, per comments in the gist)

https://gist.github.com/jhohertz/476bd616d4171649a794b8c409f8d548

**Comment by brancz** [comment]
Closing due to inactivity and not a problem with the prometheus operator in the first place.

**Comment by demisx** [comment]
@brancz Can you please elaborate on how it’s not a problem with the Prometheus operator? I am using 8.0.0 helm chart and can’t seem to get etcd scraped in kops `v1.14` cluster. Not sure if it still requires these massive workarounds https://github.com/coreos/prometheus-operator/issues/2207.  Thank you.

**Comment by brancz** [comment]
The helm chart does way more than the Prometheus operator. The Prometheus operator provides APIs and general frameworks. This is concretely about what and how the helm chart works and we neither use nor maintain the helm charts.
